### PR TITLE
KAFKA-14491: [19/N] Combine versioned store RocksDB instances into one

### PR DESCRIPTION
**Pull request by vcrfxia**
The RocksDB-based versioned store implementation introduced in https://github.com/apache/kafka/pull/13188 currently uses two physical RocksDB instances per store instance: one for the "latest value store" and another for the "segments store." This PR combines those two RocksDB instances into one by representing the latest value store as a special "reserved" segment within the segments store. This reserved segment has segment ID -1, is never expired, and is not included in the regular `Segments` methods for getting or creating segments, but is represented in the physical RocksDB instance the same way as any other segment.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 1]
```diff
@@ -821,4 +823,20 @@ public Options getOptions() {
...
+
+    /**
+     * Same as {@link Bytes#increment(Bytes)} but {@code null} is returned instead of throwing
+     * {@code IndexOutOfBoundsException} in the event of overflow.
+     *
+     * @param input bytes to increment
+     * @return A new copy of the incremented byte array, or {@code null} if incrementing would
+     *         result in overflow.
+     */
+    static Bytes incrementWithoutOverflow(final Bytes input) {
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:835]: Now that negative segment ID is allowed (only for reserved segments within `LogicalKeyValueSegments`), it's possible that a valid segment ID will overflow when incremented, and we want to handle this gracefully instead of throwing an exception when it happens.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegments.java [hunk 2]
```diff
@@ -16,15 +16,33 @@
...
+ * the fact that both types of segments share the same physical RocksDB instance.
+ * To create a reserved segment, use {@link #createReservedSegment(long, String)} instead.
+ */
 public class LogicalKeyValueSegments extends AbstractSegments<LogicalKeyValueSegment> {
 
     private final RocksDBMetricsRecorder metricsRecorder;
     private final RocksDBStore physicalStore;
 
+    // reserved segments do not expire, and are tracked here separately from regular segments
+    private final Map<Long, LogicalKeyValueSegment> reservedSegments = new HashMap<>();
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegments.java:44]: Only one reserved segment is needed for the versioned store implementation (used for the latest value store) but I've chosen to implement support for reserved segments more generally in this class. If we think this is unnecessarily complex, I can update this to be a single `Optional<LogicalKeyValueSegment>` (or equivalent) instead.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 3]
```diff
@@ -124,7 +124,6 @@ public class RocksDBStore implements KeyValueStore<Bytes, byte[]>, BatchWritingS
     protected Position position;
     private OffsetCheckpoint positionCheckpoint;
 
-    // VisibleForTesting
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:127]: This cleanup is unrelated to the changes in this PR -- I happened to notice that this method is labeled as visible for testing but is actually called in a number of places from production code, so I've removed the misleading comment.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegment.java [hunk 4]
```diff
@@ -176,7 +184,7 @@ public synchronized KeyValueIterator<Bytes, byte[]> range(final Bytes from, fina
         // with empty bytes from the returned iterator. this filtering is accomplished by
         // passing the prefix filter into StrippedPrefixKeyValueIteratorAdapter().
         final Bytes toBound = to == null
-            ? Bytes.increment(prefixKeyFormatter.getPrefix())
+            ? incrementWithoutOverflow(prefixKeyFormatter.getPrefix())
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegment.java:187]: Do we need `incrementWithoutOverflow` here, of could we just use `null`?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegmentTest.java [hunk 5]
```diff
@@ -172,30 +227,77 @@ public void shouldReturnValuesOnRange() {
...
+        try (final KeyValueIterator<Bytes, byte[]> iterator = segment1.range(new Bytes(serializeBytes("0")), null)) {
+            final LinkedList<KeyValue<String, String>> expectedContents = new LinkedList<>();
+            expectedContents.add(kv0);
+            expectedContents.add(kv1);
+            expectedContents.add(kv2);
+            assertEquals(expectedContents, getDeserializedList(iterator));
+        }
+
+        // null upper and lower bounds
+        try (final KeyValueIterator<Bytes, byte[]> iterator = segment1.range(new Bytes(serializeBytes("0")), null)) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegmentTest.java]: Should this not be `range(null, null)` ?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegmentTest.java [hunk 6]
```diff
@@ -204,7 +306,7 @@ public void shouldReturnValuesOnRange() {
         }
 
         // null upper and lower bounds
-        try (final KeyValueIterator<Bytes, byte[]> iterator = segment2.range(new Bytes(serializeBytes("0")), null)) {
+        try (final KeyValueIterator<Bytes, byte[]> iterator = negativeIdSegment.range(new Bytes(serializeBytes("0")), null)) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegmentTest.java]: as above: `range(null, null)` ?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegmentsTest.java [hunk 7]
```diff
@@ -101,10 +105,34 @@ public void shouldNotCreateSegmentThatIsAlreadyExpired() {
...
+    public void shouldNotCreateReservedSegmentFromRegularMethod() {
+        assertThrows(IllegalArgumentException.class, () -> segments.getOrCreateSegmentIfLive(-1, context, 0));
+        assertThrows(IllegalArgumentException.class, () -> segments.getOrCreateSegment(-1, context));
+    }
+
     @Test
     public void shouldCleanupSegmentsThatHaveExpired() {
         final LogicalKeyValueSegment segment1 = segments.getOrCreateSegmentIfLive(0, context, 0);
-        final LogicalKeyValueSegment segment2 = segments.getOrCreateSegmentIfLive(0, context, SEGMENT_INTERVAL * 2L);
+        final LogicalKeyValueSegment segment2 = segments.getOrCreateSegmentIfLive(2, context, SEGMENT_INTERVAL * 2L);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegmentsTest.java:135]: Nice side fix.

**Review by mjsax — COMMENTED** [review]
Overall LGTM.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegment.java (hunk 4 above)
Comment (vcrfxia) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegment.java:187]: For a logical segment, a range query with `toBound == null` means that all data in the segment (after the specified `fromBound`) should be returned. But data from other segments should NOT be returned, which is why we need to specify the incremented prefix as the upper bound for the range scan in the physical store. Except in the edge case where incrementing the prefix would cause overflow, in which case using `null` for the upper bound is correct. Thus, `incrementWithoutOverflow()` is correct here.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegmentTest.java (hunk 5 above)
Comment (vcrfxia) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegmentTest.java]: Good catch! Fixed.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegment.java (hunk 4 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegment.java:187]: > a range query with `toBound == null` means

You mean `to` here, which is the input parameter (not `toBound` which is the physical bound for the store)?

> Except in the edge case where incrementing the prefix would cause overflow, in which case using null for the upper bound is correct. 

But if we overflow, and set `toBound = null` (as returned by `incrementWithoutOverflow()`) would it not imply we scan the physical store to its end, beyond the current logical segment bound?

Comment (vcrfxia) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegment.java:187]: > You mean to here, which is the input parameter 

Yeah sorry, that's what I meant.

> But if we overflow, and set toBound = null (as returned by incrementWithoutOverflow()) would it not imply we scan the physical store to its end, beyond the current logical segment bound?

If the prefix overflows then that means that this is the last possible segment in the store, in which case scanning to the end of the physical store is the same as scanning to the end of the segment, so that's actually exactly what we want. (This only works because all prefixes are the same length -- a bunch of the other logic in this class would break too if that were not the case, though, so it's a fine assumption.)

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegment.java:187]: Ah. Thanks for clarifying -- this make sense!

**Comment by mjsax** [comment]
Same compile error as on the other PR. Re-triggered Jenkins. Let's see what happens.

**Review by mjsax — APPROVED** [review]
